File: solutions/0146-lru-cache/solution.cpp

class Node{
    public:
    Node* next;
    Node* prev;
    int val;
    int key;
    Node(){
        key = 0;
        val = 0;
        next = nullptr;
        prev = nullptr;
    }
    Node(int k,int v){
        key = k;
        val = v;
        next = nullptr;
        prev = nullptr;
    }
};


class LRUCache {
public:
    unordered_map<int,Node*> m;
    Node* head;
    Node* tail;
    int limit;

    LRUCache(int capacity) {
        limit = capacity;
        head = new Node(-1,-1);
        tail = new Node(-1,-1);
        head->next = tail;
        tail->prev = head;
    }

    int get(int key) {
        if(m.find(key)!=m.end()){
            Node* curr = m[key];
            removeNode(curr);
            insertNodeAtHead(curr);
            m[key] = curr;
            return m[key]->val;
        }
        return -1;
    }

    void insertNodeAtHead(Node* new_node){
        Node* nxt = head->next;
        head->next = new_node;
        new_node->prev = head;
        new_node->next = nxt;
        nxt->prev = new_node;
    }

    void removeNode(Node* node){
        Node* prev = node->prev;
        Node* nxt = node->next;
        node->next = nullptr;
        node->prev = nullptr;
        prev->next = nxt;
        nxt->prev = prev;
    }
    
    void put(int key, int value) {
        Node* new_node = new Node(key,value);
        if(m.find(key)!=m.end()){
            removeNode(m[key]);
        }
        m[key] = new_node;
        insertNodeAtHead(new_node);
        if(m.size()>limit){
            // remove oldest means tail of the ll
            m.erase(tail->prev->key);
            removeNode(tail->prev);
        }

    }

};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: solutions/0146-lru-cache/solution.py
class Node:
    def __init__(self,value:int,key:int):
        self.value = value
        self.key = key
        self.next = None
        self.prev = None


class LRUCache:
    def __init__(self, capacity: int):
        self.left = Node(-1,-1)
        self.right = Node(-1,-1)
        self.left.next = self.right
        self.right.prev = self.left
        self.capacity = capacity
        self.keyToNodeMap = {}

    def insertAtBeginning(self,node:Node):
        temp = self.left.next
        self.left.next = node
        node.next = temp
        temp.prev = node
        node.prev = self.left
    
    def remove(self,node:Node):
        temp = node.prev
        temp.next = node.next
        node.next.prev = temp

    def get(self, key: int) -> int:
        # refresh the key, delete and reinsert in LL
        if key in self.keyToNodeMap:
            node = self.keyToNodeMap[key]
            self.remove(node)
            self.insertAtBeginning(node)
            return node.value
        else:
            return -1
        

    def put(self, key: int, value: int) -> None:
        if key in self.keyToNodeMap:
            # remove from ll
            self.remove(self.keyToNodeMap[key])
        # insert in lru
        node = Node(value,key)
        self.keyToNodeMap[key] = node
        # insert in the beginning
        self.insertAtBeginning(node)
        # if capacity is full,remove one node
        if len(self.keyToNodeMap)>self.capacity:
            # identify last node and remove from LL
            lastnode = self.right.prev
            # remove from map
            self.keyToNodeMap.pop(lastnode.key)
            self.remove(lastnode)
        return None
        
        
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
